src/lib/include/x86_swizzle.h

```diff
@@ -2,97 +2,226 @@
 #define DISTOPIA_SWIZZLE_H
 
 #include "arch_config.h"
+#include "compiler_hints.h"
+#include "distopia_type_traits.h"
 
 #ifdef DISTOPIA_X86_SSE4_1
 
-#include <immintrin.h>
-
 #include "x86_tgintrin.h"
+#include <emmintrin.h>
+#include <immintrin.h>
 
 namespace {
 
-inline void Deinterleave3(__m128 a, __m128 b, __m128 c,
-                          __m128& x, __m128& y, __m128& z) {
-  // a = x0y0z0x1, b = y1z1x2y2, c = z2x3y3z3
-  __m128 t1 = shuffle_p<_MM_SHUFFLE(2,1,3,2)>(b, c);
-  __m128 t2 = shuffle_p<_MM_SHUFFLE(1,0,2,1)>(a, b);
+// shuffle first element of __m128 to end and shunt everything left by one
+inline __m128 ShuntFirst2Last(const __m128 input) {
+  // PRE: input = abcd
+  return shuffle_p<_MM_SHUFFLE(0, 3, 2, 1)>(input, input);
+  // return bcda
+}
+
+#ifdef DISTOPIA_X86_AVX2_FMA
+
+// shuffle first element of __m256d to end and shunt everything left by one
+// NOTE requires AVX2 rather than AVX
+// NOTE is it possible to do this with a lower latency by using shuffle and
+// blend rather than the lane crossing instruction ?
+inline __m256d ShuntFirst2Last(const __m256d input) {
+  // PRE: input = abcd
+  return _mm256_permute4x64_pd(input, _MM_SHUFFLE(0, 3, 2, 1));
+  // return bcda
+}
+
+#endif // DISTOPIA_X86_AVX2_FMA
+
+// safely loads values from an array of ScalarT to a SIMD type of VectorT with
+// width of 4, checking that we dont read off the end of ScalarT.
+template <typename VectorT>
+inline VectorT SafeIdxLoad4(const VectorToScalarT<VectorT> *source,
+                            const int idx,
+                            const VectorToScalarT<VectorT> *end) {
+  static_assert(ValuesPerPack<VectorT> == 4,
+                "can only use to load into SIMD datatype of width 4");
+  VectorT tmp;
+  if (distopia_likely(source + idx + ValuesPerPack<VectorT> < end)) {
+    // load as is, no overflow, likely path
+    tmp = loadu_p<VectorT>(&source[idx]);
+  } else {
+    // load offset by one, unlikely path
+    tmp = loadu_p<VectorT>(&source[idx - 1]);
+    tmp = ShuntFirst2Last(tmp);
+  }
+  return tmp;
+}
+
+// transforms xyz coordinates from AOS to SOA
+inline void Deinterleave4x3(const __m128 a, const __m128 b, const __m128 c,
```

Comment:
A few of these functions are redundant because they're inlined in the class right?

Reply:
Yep, but I tried to keep all the x86 swizzle math in its own header and keep `vector_triple.h` platform agnostic. That way, say we support ARM in the future, we don't pollute the class with implementation specific logic.